tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,332 @@
+    "SpinSystemInfo",
+    [
+        "g",  # Numpy `float` order parameter.
+        "gs",  # Numpy `complex` array ground state wave function from exact
+        # diagonalization.
+        "gs_energy",  # Numpy `float` ground state energy from exact
+        # diagonalization.
+        "res_energy",  # Python `float` residual between the  circuit energy and
+        # the exact energy from exact diagonalization.
+        "fidelity",  # Python `float` overlap between the circuit state and the
+        # exact ground state from exact diagonalization.
+        "params"  # Dict with Python `str` keys and Numpy`float` values.
+        # Contains $M \times P parameters. Here $M$ is the number of
+        # parameters per circuit layer and $P$ the circuit depth.
+    ])
+
+
+def unique_name():
+    """Generator to generate an infinite number of unique names.
+
+    Yields:
+        Python `str` of the form "theta_<integer>"
+
+    """
+    num = 0
+    while True:
+        yield "theta_" + str(num)
+        num += 1
+
+
+def spin_system_data_set(nspins, system_name, data_dir):
+    """Download data and load the data and convert to useful data structres
+
+    Args:
+        nspins: Python `int` number of spins in the system.
+        system_name: Python `str` name of the system.
+        data_dir: Python `str` location where to store the data on disk.
+            Default is `~/tfq-datasets`
+
+    Returns:
+        A Python `lst` of Python `float` order parameters for the system
+        A Python `lst` of `namedtuple` instances containing the following
+            fields:
+        - `g`: Numpy `float` order parameter.
+        - `gs`: Numpy `complex` array ground state wave function from
+            exact diagonalization.
+        - `gs_energy`: Numpy `float` ground state energy from exact
+            diagonalization.
+        - `res_energy`: Python `float` residual between the circuit energy
+            and the exact energy from exact diagonalization.
+        - `fidelity`: Python `float` overlap between the circuit state
+            and the exact ground state from exact diagonalization.
+        - `params`: Dict with Python `str` keys and Numpy`float` values.
+            Contains $M \times P $ parameters. Here $M$ is the number of
+            parameters per circuit layer and $P$ the circuit depth.
+
+    """
+    # Set default storage location
+    if data_dir is None:
+        data_dir = os.path.expanduser("~/tfq-datasets")
+        if not os.path.exists(data_dir):
+            os.mkdir(data_dir)
+
+    # Use Keras file downloader.
+    get_file(fname='spin_systems',
+             cache_dir=data_dir,
+             cache_subdir='',
+             origin="https://storage.googleapis.com/"
+             "download.tensorflow.org/data/"
+             "quantum/spin_systems_data.zip",
+             extract=True)
+    data_path = data_dir + "/tfq_data-master/spin_systems/" + system_name \
+                + "_{}".format(nspins)
```

Comment:
This confused me a bit, the `get_file` code looks like this (line 215 to 224):
``` python
...
  if cache_dir is None:
    cache_dir = os.path.join(os.path.expanduser('~'), '.keras')
  if md5_hash is not None and file_hash is None:
    file_hash = md5_hash
    hash_algorithm = 'md5'
  datadir_base = os.path.expanduser(cache_dir)
  if not os.access(datadir_base, os.W_OK):
    datadir_base = os.path.join('/tmp', '.keras')
  datadir = os.path.join(datadir_base, cache_subdir)
  _makedirs_exist_ok(datadir)
...
```
So if the user passes path to `cache_dir` that does not exist, the data gets cached to `/tmp/.keras`, which is undesirable because then we store the data in a temporary folder. This is why I handled this with the `os.mkdir` block.